# Advent of Code 2015

___

[**Day 1**](#day1) &nbsp; &nbsp; &nbsp; [**Day 2**](#day2) &nbsp; &nbsp; &nbsp; [**Day 3**](#day3) &nbsp; &nbsp; &nbsp; [**Day 4**](#day4) &nbsp; &nbsp; &nbsp; [**Day 5**](#day5)

[**Day 6**](#day6) &nbsp; &nbsp; &nbsp; [**Day 7**](#day7) &nbsp; &nbsp; &nbsp; [**Day 8**](#day8) &nbsp; &nbsp; &nbsp; [**Day 9**](#day9) &nbsp; &nbsp; &nbsp; [**Day 10**](#day10)

[**Day 11**](#day11) &nbsp; &nbsp; [**Day 12**](#day12) &nbsp; &nbsp; [**Day 13**](#day13) &nbsp; &nbsp; [**Day 14**](#day14) &nbsp; &nbsp; [**Day 15**](#day15)

[**Day 16**](#day16) &nbsp; &nbsp; [**Day 17**](#day17) &nbsp; &nbsp; [**Day 18**](#day18) &nbsp; &nbsp; [*Day 19*](#day19) &nbsp; &nbsp; [**Day 20**](#day20)

[**Day 21**](#day21) &nbsp; &nbsp; [Day 22](#day22) &nbsp; &nbsp; [**Day 23**](#day23) &nbsp; &nbsp; [**Day 24**](#day24) &nbsp; &nbsp; [*Day 25*](#day25)

___

<a class="anchor" id="day1"></a>
## Day 1

Santa's directions in a building are given by '(' and ')' for up and down.

##### Part 1
Determine Santa's final floor after all directions.

##### Part 2
Determine the number of the direction on which Santa first enters the basement (floor -1).

In [1]:
with open('data2015/day1.txt') as f1:
    directions = f1.read()
    
directions[:10]

'(((())))()'

**Part 1**

In [2]:
directions.count('(') - directions.count(')')

232

**Part 2**

In [3]:
position = 0
floor = 0
while floor >= 0:
    floor += 1 if directions[position] == '(' else -1
    position += 1
position

1783

<a class="anchor" id="day2"></a>
## Day 2

Elves are wrapping presents.  Each present takes its surface area in paper plus the smallest side's worth of area for slack.

##### Part 1
Determine the total area needed for all the presents.

##### Part 2
Given that the ribbon needed for each package is equal to the smallest perimeter + the volume of the box (in feet). Find the total ribbon needed.

In [4]:
with open('data2015/day2.txt') as f2:
    strdims = [row.strip() for row in f2.readlines()]
    
strdims[-3:]

['21x29x14', '20x29x30', '23x11x5']

In [5]:
dims = [(int(a), int(b), int(c)) for a, b, c in [dim.split('x') for dim in strdims]]
dims[:3]

[(4, 23, 21), (22, 29, 19), (11, 4, 11)]

**Part 1**

In [6]:
total_area = 0
for a, b, c in dims:
    area = 2*a*b + 2*b*c + 2*a*c
    small2 = sorted([a, b, c])[:2]
    area += small2[0]*small2[1]
    total_area += area
total_area

1598415

**Part 2**

In [7]:
ribbon = 0
for a, b, c in dims:
    volume = a*b*c
    small2 = sorted([a, b, c])[:2]
    ribbon += volume + (2*small2[0] + 2*small2[1])
ribbon

3812909

<a class="anchor" id="day3"></a>
## Day 3

Santa is delivering presents to houses in a grid based on directions given by <>v^.

##### Part 1
Determine the total number of houses that receive at least 1 present.

##### Part 2
The instructions are to be shared by Santa (odd instructions) and RoboSanta (even instructions). Determine the total number of houses that receive at least 1 present.

In [8]:
with open('data2015/day3.txt') as f3:
    directions = f3.read()
    
directions[-5:]

'^>vv^'

In [9]:
move_dict = {'^': (0, 1), 'v': (0, -1), '>': (1, 0), '<': (-1, 0)}

**Part 1**

In [10]:
x, y = 0, 0
houses_visited = [(0, 0)]
for direction in directions:
    dx, dy = move_dict[direction]
    x, y = x + dx, y + dy
    if (x, y) not in houses_visited:
        houses_visited.append((x, y))

len(houses_visited)

2081

**Part 2**

In [11]:
xs, ys, xr, yr = 0, 0, 0, 0
houses_visited = [(0, 0)]
for i, direction in enumerate(directions):
    dx, dy = move_dict[direction]
    if i%2:
        xs, ys = xs + dx, ys + dy
        if (xs, ys) not in houses_visited:
            houses_visited.append((xs, ys))
    else:
        xr, yr = xr + dx, yr + dy
        if (xr, yr) not in houses_visited:
            houses_visited.append((xr, yr))
            
len(houses_visited)

2341

<a class="anchor" id="day4"></a>
## Day 4

##### Part 1
Using the puzzle input, find the smallest integer we can append to the input such that the string has an MD5 hash that starts with five 0's.

##### Part 2
Same as part 1, but six 0's.

In [12]:
import hashlib

# example
hashlib.md5('abc'.encode('utf-8')).hexdigest()

'900150983cd24fb0d6963f7d28e17f72'

**Part 1**

In [13]:
day4input = 'bgvyzdsv'
i = 0
while True:
    i += 1
    to_check = day4input + str(i)
    if hashlib.md5(to_check.encode('utf-8')).hexdigest()[:5] == '00000':
        print(hashlib.md5(to_check.encode('utf-8')).hexdigest())
        break

i

000004b30d481662b9cb0c105f6549b2


254575

**Part 2**

In [14]:
day4input = 'bgvyzdsv'
i = 0
while True:
    i += 1
    to_check = day4input + str(i)
    if hashlib.md5(to_check.encode('utf-8')).hexdigest()[:6] == '000000':
        print(hashlib.md5(to_check.encode('utf-8')).hexdigest())
        break

i

000000b1b64bf5eb55aad89986126953


1038736

<a class="anchor" id="day5"></a>
## Day 5

##### Part 1
A "nice" string is one that has the following properties:
1. At least three vowels
2. At least one letter twice in a row
3. Does not contain any of: 'ab', 'cd', 'pq', 'xy'

How many nice strings?

##### Part 2
Now, a "nice" string has the following properties:
1. Contains any two letters that appear at least twice without overlapping; 'xyaxy' but not 'aaa'
2. At least one letter with repeated with exactly one letter between them; 'xyx' or 'aaa'

In [15]:
with open('data2015/day5.txt') as f5:
    strings = [row.strip() for row in f5.readlines()]

strings[-3:]

['mfifrjamczjncuym', 'otmgvsykuuxrluky', 'oiuroieurpyejuvm']

In [16]:
def is_nice1(string):
    # at least three vowels
    vowels = 'aeiou'
    cond1 = sum([string.count(v) for v in vowels]) >= 3
    
    # use pairs list for cond2 and cond3
    pairs = [string[i:i+2] for i in range(len(string)-1)]
    cond2 = any([pair[0] == pair[1] for pair in pairs])
    
    bad = ['ab', 'cd', 'pq', 'xy']
    cond3 = any([b == p for b in bad for p in pairs])
    
    return cond1 and cond2 and not cond3


In [17]:
tests1 = ['ugknbfddgicrmopn',  # nice
         'aaa',               # nice
         'jchzalrnumimnmhp',  # naughty - no double
         'haegwjzuvuyypxyu',  # naughty - has xy
         'dvszwmarrgswjxmb']  # naughty - one vowel

for test in tests1:
    print(is_nice1(test))

True
True
False
False
False


In [18]:
part1 = sum([is_nice1(string) for string in strings])
part1

258

In [19]:
def is_nice2(string):
    pairs = [string[i:i+2] for i in range(len(string)-1)]
    cond1 = False
    for i, p1 in enumerate(pairs[:-1]):
        for j, p2 in enumerate(pairs[i+2:]):
            if p1 == p2:
                cond1 = True
    cond2 = any([string[i] == string[i+2] for i in range(len(string)-2)])
    return cond1 and cond2

is_nice2(tests1[0])

False

In [20]:
tests2 = ['qjhvhtzxzqqjkmpb',  # nice 
          'xxyxx',             # nice
          'uurcxstgmygtbstg',  # naughty
          'ieodomkazucvgmuy']  # naughty

for test in tests2:
    print(is_nice2(test))

True
True
False
False


In [21]:
part2 = sum([is_nice2(string) for string in strings])
part2

53

<a class="anchor" id="day6"></a>
## Day 6

The million-light display.

##### Part 1
All lights begin off.  Following all of Santa's instructions, what is the final number of lights that are on?

##### Part 2
Instead, start with all lights at a brightness of 0. Interpret 'turn on' to mean increase brightness by 1, 'turn off' to mean decrease by 1 to a min of 0, and 'toggle' to mean increase brightness by 2. Determine the sum of all brightnesses at the end.

In [22]:
import numpy as np

with open('data2015/day6.txt') as f6:
    instructions = [row.strip() for row in f6.readlines()]
    
instructions[-3:]

['toggle 534,948 through 599,968',
 'turn on 522,730 through 968,950',
 'turn off 102,229 through 674,529']

In [23]:
clean_insts = []
for inst in instructions:
    ci = inst.split(' ')
    ci.remove('through')
    if 'turn' in ci:
        ci.remove('turn')
    ci = ci[:1] + [int(x) for x in ci[1].split(',')] + [int(x) for x in ci[2].split(',')] 
    clean_insts.append(ci)
clean_insts[:8]

[['on', 887, 9, 959, 629],
 ['on', 454, 398, 844, 448],
 ['off', 539, 243, 559, 965],
 ['off', 370, 819, 676, 868],
 ['off', 145, 40, 370, 997],
 ['off', 301, 3, 808, 453],
 ['on', 351, 678, 951, 908],
 ['toggle', 720, 196, 897, 994]]

In [24]:
lights1 = np.ones(shape=(1000, 1000))*(-1)
for ci in clean_insts:
    s, x0, y0, x1, y1 = ci
    if s == 'on':
        lights1[x0:x1+1, y0:y1+1] = 1
    elif s == 'off':
        lights1[x0:x1+1, y0:y1+1] = -1
    elif s == 'toggle':
        lights1[x0:x1+1, y0:y1+1] *= -1

part1 = sum(sum(lights1 == 1))
part1

377891

In [25]:
lights2 = np.zeros(shape=(1000, 1000))
for ci in clean_insts:
    s, x0, y0, x1, y1 = ci
    if s == 'on':
        lights2[x0:x1+1, y0:y1+1] += 1
    elif s == 'off':
        lights2[x0:x1+1, y0:y1+1] -= 1
        lights2 = np.where(lights2 < 0, 0, lights2)
    elif s == 'toggle':
        lights2[x0:x1+1, y0:y1+1] += 2

part2 = sum(sum(lights2))
part2

14110788.0

<a class="anchor" id="day7"></a>
## Day 7

Bitwise operators with circuits.

##### Part 1
After following all of the instructions, what is the signal provided to wire 'a'?

##### Part 2
Now override the signal into wire 'b' as the signal going to 'a' from Part 1.  Doing this (and knowing that each wire can only receive one signal, and 'b' is now already receiving a signal), reassemble the circuits and re-determine the signal going to 'a'.

In [26]:
sample = '''123 -> x
456 -> y
x AND y -> d
x OR y -> e
x LSHIFT 2 -> f
y RSHIFT 2 -> g
NOT x -> h
NOT y -> i'''

In [27]:
with open('data2015/day7.txt') as f7:
    circuits = [row.strip() for row in f7.readlines()]

circuits[-3:]

['NOT h -> i', 'NOT hn -> ho', 'he RSHIFT 5 -> hh']

In [28]:
def get_signal(wire, circuits, signals):
    
    while len(circuits) > 0:
        idx_to_remove = []
        for i, row in enumerate(circuits):
            row = row.split(' ')
            arrow_ind = row.index('->')
            if arrow_ind == 1:
                # this "if row[-1]"statement is the key to part 2 - the problem is sort of vague about the "override to b",
                #  thing but it really means that b can't get set to anything else because it is already receiving a signal,
                #  so we need to skip the 'b -> (some num)' statement that is encountered or else we just end up with the
                #  exact same signals as in part 1
                if row[-1] in signals:
                    idx_to_remove.append(i)
                    continue
                # sometimes get things like 3 -> x, other times things like y -> x
                if row[0][0] in '0123456789':
                    signals[row[-1]] = int(row[0])
                    idx_to_remove.append(i)
                else:
                    if row[0] in signals:
                        signals[row[-1]] = signals[row[0]]
                        idx_to_remove.append(i)
            elif arrow_ind == 2:
                if row[1] in signals:
                    signals[row[-1]] = (~signals[row[1]]) % 65536
                    idx_to_remove.append(i)
            elif arrow_ind == 3:
                # sometimes AND statements have a number: 1 and x -> y
                if row[1] == 'AND':
                    if (row[0][0] in '0123456789') and (row[2] in signals):
                        signals[row[-1]] = int(row[0]) & signals[row[2]]
                        idx_to_remove.append(i)
                    else:
                        if row[0] in signals and row[2] in signals:
                            signals[row[-1]] = signals[row[0]] & signals[row[2]]
                            idx_to_remove.append(i)
                elif row[1] == 'OR':
                    if row[0] in signals and row[2] in signals:
                        signals[row[-1]] = signals[row[0]] | signals[row[2]]
                        idx_to_remove.append(i)
                elif row[1] == 'LSHIFT':
                    if row[0] in signals:
                        signals[row[-1]] = signals[row[0]] << int(row[2])
                        idx_to_remove.append(i)
                elif row[1] == 'RSHIFT':
                    if row[0] in signals:
                        signals[row[-1]] = signals[row[0]] >> int(row[2])
                        idx_to_remove.append(i)
        circuits = [circuits[i] for i in range(len(circuits)) if i not in idx_to_remove]

    return signals[wire]

In [29]:
get_signal('h', sample.split('\n'), dict())

65412

In [30]:
part1 = get_signal('a', circuits, dict())
part1

16076

In [31]:
part2 = get_signal('a', circuits, {'b': part1})
part2

2797

<a class="anchor" id="day8"></a>
## Day 8

Comparing difference in length of string literals (how many actual characters typed) vs in-memory code size.

##### Part 1
What is the difference between the number of characters of code for string literals minus the number of characters in memory for the values of strings?

##### Part 2
Now, encode each string as a new string, such as "" -> "\"\"".  Then, again find the sum of the differences between the newly encoded string and the number of characters in the original.

Samples:
    
    sample1 = ''           # 2  -  0 = 2
    sample2 = 'abc'        # 5  -  3 = 2
    sample3 = 'aaa\'aaa'   # 10 -  7 = 3
    sample4 = '\x27'       # 6  -  1 = 5 (represents an apostrophe)
                           #  total  = 12

In [32]:
with open('data2015/day8.txt') as f8:
    strings = f8.read()

# chop off trailing newline
strings = strings[:-1]

In [33]:
part1 = 0
for string in strings.split('\n'):
    part1 += len(string) - len(eval(string))

part1

1333

In [34]:
part2 = 0
for string in strings.split('\n'):
    part2 += string.count('\\') + string.count('"') + 2
part2

2046

<a class="anchor" id="day9"></a>
## Day 9

Finding shortest path between different cities.

##### Part 1
Find the shortest path that visits every location exactly once.

##### Part 2
Find the longest path that visits every location exactly once.

In [35]:
sample = '''London to Dublin = 464
London to Belfast = 518
Dublin to Belfast = 141'''

In [36]:
def get_travel_dict_and_locations(text):
    travel_dict = dict()
    locations = []
    rows = text.split('\n')
    if rows[-1] == '':
        rows = rows[:-1]
    for row in rows:
        row = row.split(' ')
        if row[0] not in locations:
            locations.append(row[0])
        if row[2] not in locations:
            locations.append(row[2])
        travel_dict[(row[0], row[2])] = int(row[-1])
        travel_dict[(row[2], row[0])] = int(row[-1])

    return travel_dict, locations

get_travel_dict_and_locations(sample)

({('London', 'Dublin'): 464,
  ('Dublin', 'London'): 464,
  ('London', 'Belfast'): 518,
  ('Belfast', 'London'): 518,
  ('Dublin', 'Belfast'): 141,
  ('Belfast', 'Dublin'): 141},
 ['London', 'Dublin', 'Belfast'])

In [37]:
import itertools

# this function works for parts 1 and 2; which can take on either 'shortest' or 'longest'
def get_best_route(travel_dict, locations, which='shortest'):
    routes = list(itertools.permutations(locations))
    best_route = ()
    best_length = 100000000 if which == 'shortest' else 0
    for route in routes:
        route_length = 0
        for i in range(len(route)-1):
            route_length += travel_dict[(route[i], route[i+1])]
        if which == 'shortest':
            if route_length < best_length:
                best_route = route
                best_length = route_length
        elif which == 'longest':
            if route_length > best_length:
                best_route = route
                best_length = route_length
    return best_route, best_length

In [38]:
travel_dict, locations = get_travel_dict_and_locations(sample)
best_route, best_length = get_best_route(travel_dict, locations, which='shortest')
best_route, best_length

(('London', 'Dublin', 'Belfast'), 605)

In [39]:
with open('data2015/day9.txt') as f9:
    distances = f9.read()
    
travel_dict, locations = get_travel_dict_and_locations(distances)

In [40]:
# Part 1
best_route, best_length = get_best_route(travel_dict, locations, which='shortest')
best_route, best_length

(('Tambi',
  'Arbre',
  'Snowdin',
  'AlphaCentauri',
  'Tristram',
  'Straylight',
  'Faerun',
  'Norrath'),
 251)

In [41]:
# Part 2
best_route, best_length = get_best_route(travel_dict, locations, which='longest')
best_route, best_length

(('Tristram',
  'Faerun',
  'Arbre',
  'Straylight',
  'AlphaCentauri',
  'Norrath',
  'Tambi',
  'Snowdin'),
 898)

<a class="anchor" id="day10"></a>
## Day 10

Look and say sequence

##### Part 1
Apply the look and say process to the given input 40 times. What is the length of the result?

##### Part 2
Apply the look and say process to the given input 50 times. What is the length of the result?

In [42]:
day10 = '3113322113'

def dolooknsay(nums):
    newstr = ''
    i = 0
    while i < len(nums):
        current = nums[i]
        j = 1
        if i + j >= len(nums):
            newstr += '1' + current
            break
        while nums[i + j] == current:
            j += 1
        newstr += str(j) + current
        i += j
    return newstr
dolooknsay(day10)

'132123222113'

In [43]:
nums = day10[:]
for i in range(40):
    nums = dolooknsay(nums)
part1 = len(nums)
part1

329356

In [44]:
nums = day10[:]
for i in range(50):
    nums = dolooknsay(nums)
part2 = len(nums)
part2

4666278

<a class="anchor" id="day11"></a>
## Day 11

Password policies.  All passwords are 8 lowercase letters.  

##### Part 1
The rules are:
1. password must contain one increasing straight of at at least three letters, like 'abc' or 'hijk'
2. password can't contain letters i, o, or l.
3. password must contain two non-overlapping pairs of letters, like 'aabb', 'axxbyy', but not 'axxaxx' or 'bbb'.
Given Santa's current password (puzzle input), determine the next valid password alphabetically (from aax, check aay, aaz, aba, ...)

In [45]:
alphabet = 'abcdefghijklmnopqrstuvwxyz'
day11 = 'cqjxjnds'

In [46]:
def condition1(pw):
    for i in range(6):
        if pw[i:i+3] in alphabet:
            return True
    return False

condition1('amvrstyw')

True

In [47]:
def condition2(pw):
    return not any([bad in pw for bad in ['i', 'o', 'l']])

condition2('akjhorer')

False

In [48]:
def condition3(pw):
    pairs_found = 0
    i = 0
    while pairs_found < 2 and i < 7:
        if pw[i] == pw[i+1]:
            pairs_found += 1
            i += 2
        else:
            i += 1
    return pairs_found == 2

In [49]:
def get_next_pw(pw):
    for i in range(7, -1, -1):
        if pw[i] != 'z':
            return pw[:i] + alphabet[alphabet.index(pw[i]) + 1] + 'a'*(7-i)
    
get_next_pw('abzztzzz')

'abzzuaaa'

In [50]:
def get_next_good_pw(curr_pw):
    pw = get_next_pw(curr_pw)
    while True:
        if all([condition1(pw), condition2(pw), condition3(pw)]):
            break
        else:
            pw = get_next_pw(pw)
    return pw

In [51]:
part1 = get_next_good_pw(day11)
part1

'cqjxxyzz'

In [52]:
part2 = get_next_good_pw(part1)
part2

'cqkaabcc'

<a class="anchor" id="day12"></a>
## Day 12 

JSON

##### Part 1
What is the sum of all of the numbers in the document? (note - this is much simpler interpreting the data as a raw text file rather than interpreting it as JSON data)

##### Part 2
Now, find the sum of all numbers in the document *that are not part of a dict with a value of 'red'*.

In [53]:
with open('data2015/day12.txt') as f12:
    doc = f12.read().strip('\n')

doc[-50:]

'{"a":138},"b":118,"g":"green","f":0,"i":"violet"}}'

In [54]:
# sample = '{"a":138},"b":-118,"g":"green","f":0,"i":"violet"}}'
# doc = sample

In [55]:
part1 = 0
nums = '-0123456789'
i = 0
while i < len(doc):
    if doc[i] in nums:
        newstr = doc[i]
        while doc[i+1] in nums:
            newstr += doc[i+1]
            i += 1
        part1 += int(newstr)
    i += 1
part1

111754

In [56]:
import json

with open('data2015/day12.txt') as json_file:
    data = json.load(json_file)
data

{'e': [[{'e': 86,
    'c': 23,
    'a': {'a': [120, 169, 'green', 'red', 'orange'], 'b': 'red'},
    'g': 'yellow',
    'b': ['yellow'],
    'd': 'red',
    'f': -19},
   {'e': -47,
    'a': [2],
    'd': {'a': 'violet'},
    'c': 'green',
    'h': 'orange',
    'b': {'e': 59,
     'a': 'yellow',
     'd': 'green',
     'c': 47,
     'h': 'red',
     'b': 'blue',
     'g': 'orange',
     'f': ['violet', 43, 168, 78]},
    'g': 'orange',
    'f': [{'e': [82, -41, 2, 'red', 'violet', 'orange', 'yellow'],
      'c': 'green',
      'a': 77,
      'g': 'orange',
      'b': 147,
      'd': 49,
      'f': 'blue'},
     -1,
     142,
     136,
     ['green', 'red', 166, -21],
     'blue',
     'orange',
     {'a': 38}]},
   'orange',
   'yellow'],
  'green',
  -22,
  [37,
   [4,
    -40,
    ['red',
     'yellow',
     ['yellow',
      177,
      'red',
      'blue',
      139,
      [55, 13, 'yellow', 'violet', -21, 140, 'yellow', 117],
      'blue',
      'blue',
      106],
     'blue',
   

In [57]:
type(data)

dict

In [58]:
# recursively score each chunk from the whole document down
def score_object(obj):
    
    def score_dict(D):
        score = 0
        for key in D:
            val = D[key]
            if val == 'red':
                return 0
            if type(val) == dict:
                score += score_dict(val)
            elif type(val) == list:
                score += score_list(val)
            else:
                score += score_other(val)
        return score
    
    def score_list(L):
        score = 0
        for obj in L:
            if type(obj) == dict:
                score += score_dict(obj)
            elif type(obj) == list:
                score += score_list(obj)
            else:
                score += score_other(obj)
        return score
    
    def score_other(other):
        return other if type(other) == int else 0
    
    score = 0
    if type(obj) == dict:
        return score_dict(obj)
    elif type(obj) == list:
        return score_list(obj)
    else:
        return score_other(obj)
    
    return score
            
part2 = score_object(data)
part2

65402

<a class="anchor" id="day13"></a>
## Day 13

Seating arrangements

##### Part 1


In [59]:
sample13 = '''Alice would gain 54 happiness units by sitting next to Bob.
Alice would lose 79 happiness units by sitting next to Carol.
Alice would lose 2 happiness units by sitting next to David.
Bob would gain 83 happiness units by sitting next to Alice.
Bob would lose 7 happiness units by sitting next to Carol.
Bob would lose 63 happiness units by sitting next to David.
Carol would lose 62 happiness units by sitting next to Alice.
Carol would gain 60 happiness units by sitting next to Bob.
Carol would gain 55 happiness units by sitting next to David.
David would gain 46 happiness units by sitting next to Alice.
David would lose 7 happiness units by sitting next to Bob.
David would gain 41 happiness units by sitting next to Carol.
'''

def get_happy_dict(text):
    hd = dict()
    for row in text.split('\n'):
        if row != '':
            row = row.split(' ')
            p1, p2, val = row[0], row[-1][:-1], int(row[3]) if row[2] == 'gain' else -int(row[3])
            if (p2, p1) not in hd:
                hd[(p1, p2)] = val
            else:
                hd[(p2, p1)] += val
                
    # add reversed pairs with same value to dict for easier lookup later
    revs = {(k2, k1): v for (k1, k2), v in hd.items()}
    hd.update(revs)
    
    return hd

hd = get_happy_dict(sample13)
hd

{('Alice', 'Bob'): 137,
 ('Alice', 'Carol'): -141,
 ('Alice', 'David'): 44,
 ('Bob', 'Carol'): 53,
 ('Bob', 'David'): -70,
 ('Carol', 'David'): 96,
 ('Bob', 'Alice'): 137,
 ('Carol', 'Alice'): -141,
 ('David', 'Alice'): 44,
 ('Carol', 'Bob'): 53,
 ('David', 'Bob'): -70,
 ('David', 'Carol'): 96}

In [60]:
import itertools
def find_best_arrangement(happy_dict, include_you=False):
    
    # get the unique people from the happy dict
    people = list(set([person for pair in happy_dict.keys() for person in pair]))
    
    if include_you:
        happy_dict.update({('you', person): 0 for person in people})
        happy_dict.update({(person, 'you'): 0 for person in people})
        people.append('you')
        
    # to get all permuations *unique up to rotation*, find permutations of first n-1 items, then append nth to all of them
    orderings = list(itertools.permutations(people[:-1]))
    orderings = [people[-1:] + list(order) for order in orderings]
    
    # score each arrangement
    best_score, best_order = 0, []
    for order in orderings:
        score = 0
        for i in range(len(people)-1):
            score += happy_dict[(order[i], order[i+1])]
        score += happy_dict[(order[-1], order[0])]
        if score > best_score:
            best_score = score
            best_order = order
            
    return best_score, best_order

find_best_arrangement(hd)

(330, ['Bob', 'Carol', 'David', 'Alice'])

In [61]:
with open('data2015/day13.txt') as f13:
    happy_dict = get_happy_dict(f13.read())

In [62]:
part1 = find_best_arrangement(happy_dict, include_you=False)
part1

(709, ['Bob', 'Mallory', 'Eric', 'Carol', 'Frank', 'David', 'George', 'Alice'])

In [63]:
part2 = find_best_arrangement(happy_dict, include_you=True)
part2

(668,
 ['you',
  'Frank',
  'Carol',
  'Eric',
  'Mallory',
  'Bob',
  'Alice',
  'George',
  'David'])

<a class="anchor" id="day14"></a>
## Day 14

Reindeer Race

##### Part 1
After 2503 seconds, what is the distance traveled by the winning reindeer?

##### Part 2
Now each second the reindeer in the lead gets one point (or all that are tied for the lead). What is the most number of points any reindeer has after 2503 seconds?

*Wrote a Reindeer class for this one in day14reindeer.py*

In [64]:
from data2015.day14reindeer import Reindeer

In [65]:
sample = '''Comet can fly 14 km/s for 10 seconds, but then must rest for 127 seconds.
Dancer can fly 16 km/s for 11 seconds, but then must rest for 162 seconds.
'''

def initialize_race(text):
    alldeer = []
    for row in text.split('\n')[:-1]:
        row = row.split(' ')
        name, speed, mft, mrt = row[0], int(row[3]), int(row[6]), int(row[-2])
        alldeer.append(Reindeer(name, speed, mft, mrt))
    return alldeer

alldeer = initialize_race(sample)
for t in range(1000):
    for deer in alldeer:
        deer.do_second()

for deer in alldeer:
    print(deer.name, deer.distance)
    

Comet 1120
Dancer 1056


In [66]:
actual14 = '''Dancer can fly 27 km/s for 5 seconds, but then must rest for 132 seconds.
Cupid can fly 22 km/s for 2 seconds, but then must rest for 41 seconds.
Rudolph can fly 11 km/s for 5 seconds, but then must rest for 48 seconds.
Donner can fly 28 km/s for 5 seconds, but then must rest for 134 seconds.
Dasher can fly 4 km/s for 16 seconds, but then must rest for 55 seconds.
Blitzen can fly 14 km/s for 3 seconds, but then must rest for 38 seconds.
Prancer can fly 3 km/s for 21 seconds, but then must rest for 40 seconds.
Comet can fly 18 km/s for 6 seconds, but then must rest for 103 seconds.
Vixen can fly 18 km/s for 5 seconds, but then must rest for 84 seconds.
'''

In [67]:
alldeer = initialize_race(actual14)
for t in range(2503):
    for deer in alldeer:
        deer.do_second()

for deer in alldeer:
    print(deer.name, deer.distance)
print()
part1 = print(max(deer.distance for deer in alldeer))
part1

Dancer 2565
Cupid 2596
Rudolph 2640
Donner 2548
Dasher 2304
Blitzen 2590
Prancer 2589
Comet 2484
Vixen 2610

2640


In [68]:
alldeer = initialize_race(actual14)
for t in range(2503):
    for deer in alldeer:
        deer.do_second()
    best_dist = max(deer.distance for deer in alldeer)
    for deer in alldeer:
        if deer.distance == best_dist:
            deer.get_point()
        
        
for deer in alldeer:
    print(deer.name, deer.points)
print()
part2 = max(deer.points for deer in alldeer)
part2

Dancer 164
Cupid 46
Rudolph 647
Donner 1102
Dasher 0
Blitzen 6
Prancer 176
Comet 213
Vixen 360



1102

<a class="anchor" id="day15"></a>
## Day 15

Each cookie ingredient has different properties that, when mixed in different portions, result in cookies with different scores. Each cookie gets 100 total parts in its ingredients (for the sample, might have 31 parts butterscotch and 69 parts cinnamon).

##### Part 1
Find the maximum possible score for a cook with the given ingredients/properties. Score by taking the dot product of amounts and property values for each property, then find those dot products' scalar product.  Any dot product that results in a negative value should be set to zero (so that the overall scalar product score for that cookie will be 0). Do not use calories for Part 1.
    
##### Part 2
Now, find the optimal score for a cookie that has 500 calories.

(to implement code for Part1 vs. Part2, just modify the "counting_calories" boolean variable)
    

In [69]:
sample = {'Butterscotch': {'capacity': -1, 'durability': -2, 'flavor': 6, 'texture': 3, 'calories': 8}, 
          'Cinnamon': {'capacity': 2, 'durability': 3, 'flavor': -2, 'texture': -1, 'calories': 3}}

ings = list(sample.keys())
props = list(sample['Cinnamon'].keys())
print(ings)
final_prods = []
for i in range(1, 100):
    a = i
    b = 100 - i
    results = [a*sample[ings[0]][prop] + b*sample[ings[1]][prop] for prop in props if prop != 'calories']
    results = [result if result > 0 else 0 for result in results]
    prod = 1
    for result in results:
        prod *= result
    final_prods.append((prod, i))

max(final_prods)
    

['Butterscotch', 'Cinnamon']


(62842880, 44)

In [70]:
actual15 = {'Sprinkles': {'capacity': 5, 'durability': -1, 'flavor': 0, 'texture': 0, 'calories': 5},
            'PeanutButter': {'capacity': -1, 'durability': 3, 'flavor': 0, 'texture': 0, 'calories': 1},
            'Frosting': {'capacity': 0, 'durability': -1, 'flavor': 4, 'texture': 0, 'calories': 6},
            'Sugar': {'capacity': -1, 'durability': 0, 'flavor': 0, 'texture': 2, 'calories': 8}}

ings = list(actual15.keys())
props = list(actual15[ings[0]].keys())
final_prods = []

# set this to False for Part1, True for Part2
counting_calories = True

for i in range(1, 98):
    for j in range(i+1, 99):
        for k in range(j+1, 100):
            a = i
            b = j - i
            c = k - j
            d = 100 - k
            if counting_calories:
                cal = 'calories'
                calories = a*actual15[ings[0]][cal] + b*actual15[ings[1]][cal] + c*actual15[ings[2]][cal] + d*actual15[ings[3]][cal]
                if calories != 500:
                    continue
            results = [a*actual15[ings[0]][prop] + b*actual15[ings[1]][prop] + c*actual15[ings[2]][prop] + d*actual15[ings[3]][prop]
                       for prop in props if prop != 'calories']
            results = [result if result > 0 else 0 for result in results]
            prod = 1
            for result in results:
                prod *= result
            final_prods.append((prod, (a, b, c, d)))

max(final_prods)

(11171160, (27, 27, 15, 31))

<a class="anchor" id="day16"></a>
## Day 16

Find the correct Aunt Sue out of 500 by using what you know about all of them.

##### Part 1
Given what you remember about each of the 500 Aunt Sues, determine which one must be the one that sent you the present (the correct sue).

##### Part 2
Now, it turns out that the values for 'cats' and 'trees' represent lower bounds for that value and 'pomeranians' and 'goldfish' represent upper bounds for that value from the correct sue. Now determine the number sue that sent you the present.


In [71]:
correct_sue = '''children: 3
cats: 7
samoyeds: 2
pomeranians: 3
akitas: 0
vizslas: 0
goldfish: 5
trees: 3
cars: 2
perfumes: 1'''

correct_sue_dict = {row.split(': ')[0]: int(row.split(': ')[1]) for row in correct_sue.split('\n')}
correct_sue_dict
columns = list(correct_sue_dict.keys())
print(correct_sue_dict, columns)

{'children': 3, 'cats': 7, 'samoyeds': 2, 'pomeranians': 3, 'akitas': 0, 'vizslas': 0, 'goldfish': 5, 'trees': 3, 'cars': 2, 'perfumes': 1} ['children', 'cats', 'samoyeds', 'pomeranians', 'akitas', 'vizslas', 'goldfish', 'trees', 'cars', 'perfumes']


In [72]:
with open('data2015/day16.txt') as f16:
    sues = [row.strip() for row in f16.readlines()]

sues[:5]

['Sue 1: children: 1, cars: 8, vizslas: 7',
 'Sue 2: akitas: 10, perfumes: 10, children: 5',
 'Sue 3: cars: 5, pomeranians: 4, vizslas: 1',
 'Sue 4: goldfish: 5, children: 8, perfumes: 3',
 'Sue 5: vizslas: 2, akitas: 7, perfumes: 6']

In [73]:
import pandas as pd
import numpy as np

sue_df = pd.DataFrame(columns=columns)

# turn each row into a properly typed dict, fill in nans in the dict, add the row(as a dict) to the overall sue_df
for i, sue in enumerate(sues):
    
    # don't need the sue's number, add a ',' to end of last number for consistency with other numbers
    sue = sue.split(' ')[2:]
    sue[-1] += ','
    sue = [sue[i][:-1] if i%2==0 else int(sue[i][:-1]) for i in range(len(sue))]
    sue_dict = {sue[i]: sue[i+1] for i in range(0, len(sue), 2)}
    for col in columns:
        if col not in sue_dict:
            sue_dict[col] = np.nan
    sue_df = sue_df.append(sue_dict, ignore_index=True)
    
# sues are numbered 1-500, not 0-499
sue_df.index = [i + 1 for i in sue_df.index]

In [74]:
sue_df.head()

children  cats  samoyeds  pomeranians  akitas  vizslas  goldfish  trees  \
1       1.0   NaN       NaN          NaN     NaN      7.0       NaN    NaN   
2       5.0   NaN       NaN          NaN    10.0      NaN       NaN    NaN   
3       NaN   NaN       NaN          4.0     NaN      1.0       NaN    NaN   
4       8.0   NaN       NaN          NaN     NaN      NaN       5.0    NaN   
5       NaN   NaN       NaN          NaN     7.0      2.0       NaN    NaN   

   cars  perfumes  
1   8.0       NaN  
2   NaN      10.0  
3   5.0       NaN  
4   NaN       3.0  
5   NaN       6.0

In [75]:
sue_df2 = sue_df.copy()
for key in correct_sue_dict:
    print(key)
    mask1 = sue_df2[key] == correct_sue_dict[key]
    mask2 = sue_df2[key].isnull()
    sue_df2 = sue_df2[mask1 | mask2]

children
cats
samoyeds
pomeranians
akitas
vizslas
goldfish
trees
cars
perfumes


In [76]:
sue_df2

children  cats  samoyeds  pomeranians  akitas  vizslas  goldfish  trees  \
213       3.0   NaN       NaN          NaN     NaN      0.0       5.0    NaN   

     cars  perfumes  
213   NaN       NaN

In [77]:
sue_df3 = sue_df.copy()
for key in correct_sue_dict:
    if key in ['cats', 'trees']:
        mask1 = sue_df3[key] > correct_sue_dict[key]
    elif key in ['pomeranians', 'goldfish']:
        mask1 = sue_df3[key] < correct_sue_dict[key]
    else:
        mask1 = sue_df3[key] == correct_sue_dict[key]
    mask2 = sue_df3[key].isnull()
    sue_df3 = sue_df3[mask1 | mask2]
        

In [78]:
sue_df3

children  cats  samoyeds  pomeranians  akitas  vizslas  goldfish  trees  \
323       NaN   NaN       NaN          NaN     NaN      NaN       0.0    6.0   

     cars  perfumes  
323   NaN       1.0

<a class="anchor" id="day17"></a>
## Day 17

The knapsack problem

##### Part 1
Given containers of the sizes shown in the input, how many ways can we hold exactly 150 liters?

In [79]:
sample = [20, 15, 10, 5, 5]

def find_group(goal, so_far, available):
    #print(so_far, available)
    if so_far == goal:
        return 1
    else:
        total = 0
        for i, av in enumerate(available):
            if so_far + av <= goal:
                total += find_group(goal, so_far+av, available[i+1:])
        return total
    
find_group(25, 0, sample)

4

In [80]:
actual17 = sorted([50, 44, 11, 49, 42, 46, 18, 32, 26, 40, 21, 7, 18, 43, 10, 47, 36, 24, 22, 40])[::-1]
part1 = find_group(150, 0, actual17)
part1

654

In [81]:
# by inspection from the print statment in the 'if' block, 4 is the fewest possible containers

def find_group2(goal, so_far_list, available):
    if so_far_list != [] and sum(so_far_list) == goal:
        #print(len(so_far_list))
        return 1 if len(so_far_list) == 4 else 0
    else:
        total = 0
        for i, av in enumerate(available):
            total_so_far = 0 if so_far_list == [] else sum(so_far_list)
            if total_so_far + av <= goal:
                total += find_group2(goal, so_far_list + [av], available[i+1:])
        return total

In [82]:
part2 = find_group2(150, [], actual17)
part2

57

<a class="anchor" id="day18"></a>
## Day 18

Light show round 2 - a gif now

In [83]:
import numpy as np

with open('data2015/day18.txt') as f18:
    init_grid = [row.strip() for row in f18.readlines()]

start_grid = np.zeros((100, 100))
for i in range(100):
    for j in range(100):
        if init_grid[i][j] == '#':
            start_grid[i, j] = 1
start_grid[:10, :10]

array([[1., 0., 0., 0., 1., 1., 0., 0., 0., 0.],
       [1., 1., 1., 1., 0., 0., 1., 0., 1., 0.],
       [0., 0., 0., 1., 0., 0., 1., 1., 0., 0.],
       [0., 1., 0., 0., 0., 1., 1., 0., 0., 0.],
       [0., 0., 1., 1., 0., 1., 1., 1., 1., 1.],
       [0., 1., 1., 1., 0., 1., 1., 1., 0., 1.],
       [1., 0., 1., 1., 1., 1., 1., 0., 0., 1.],
       [0., 0., 0., 0., 1., 0., 0., 0., 1., 1.],
       [1., 1., 0., 1., 1., 0., 0., 0., 1., 0.],
       [1., 0., 1., 0., 0., 1., 0., 0., 0., 0.]])

In [84]:
def get_surround_on_count(grid, i, j):
    poss_locs = [[i+1, j+1],
                 [i+1, j+0],
                 [i+1, j-1],
                 [i+0, j+1],
                 [i+0, j-1],
                 [i-1, j+1],
                 [i-1, j+0],
                 [i-1, j-1]]
    locs = [loc for loc in poss_locs if 0 <= loc[0] <= 99 and 0 <= loc[1] <= 99]
    count = sum([grid[loc[0], loc[1]] for loc in locs])
    return count

get_surround_on_count(start_grid, 3, 0)

1.0

In [85]:
# to get the answer to Part1, set corners_on to False, for Part2, set corners_on to True 

grid = start_grid.copy()

corners_on = False

if corners_on:
    grid[0, 0], grid[0, 99], grid[99, 0], grid[99, 99] = 1, 1, 1, 1

for step in range(100):
    newgrid = np.zeros((100, 100))
    for i in range(100):
        for j in range(100):
            count = get_surround_on_count(grid, i, j)
            newgrid[i, j] = 1 if ((grid[i, j] == 1 and count in [2, 3]) or (grid[i, j] == 0 and count == 3)) else 0
            
    if corners_on:
        newgrid[0, 0], newgrid[0, 99], newgrid[99, 0], newgrid[99, 99] = 1, 1, 1, 1
    
    grid = newgrid
    
sum(sum(grid))            

814.0

<a class="anchor" id="day19"></a>
## Day 19

Reindeer chemistry

##### Part 1
Given the list of chemical replacements and the medicine molecule used to calibrate the machine, how many unique molecules can be created when performing a single replacement?

In [86]:
sample_repl = '''H => HO
H => OH
O => HH'''

sample_med = 'HOH'

In [87]:
def get_rep_tuples(text):
    rep_tuples = []
    for row in text.split('\n'):
        row = row.split(' => ')
        rep_tuples.append((row[0], row[1]))
    return rep_tuples
get_rep_tuples(sample_repl)

[('H', 'HO'), ('H', 'OH'), ('O', 'HH')]

In [88]:
def get_distinct_new_mol(rep_tuples, medicine):
    new_list = []
    i = 0
    rep_inputs = [rep[0] for rep in rep_tuples]
    while i < len(medicine):
        
        # parts of the medicine we're replacing are one or two chars long, but the incoming replacement
        #  parts have various lengths
        
        if medicine[i] in rep_inputs:
            reps = [rep for rep in rep_tuples if rep[0] == medicine[i]]
            for rep in reps:
                new = medicine[:i] + rep[1] + medicine[i+1:]
                if new not in new_list:
                    new_list.append(new)
        
        elif medicine[i:i+2] in rep_inputs:
            reps = [rep for rep in rep_tuples if rep[0] == medicine[i:i+2]]
            for rep in reps:
                new = medicine[:i] + rep[1] + medicine[i+2:]
                if new not in new_list:
                    new_list.append(new)
            
        i += 1
        
    return new_list

out = get_distinct_new_mol(get_rep_tuples(sample_repl), 'HOHOHO')
len(out), out

(7,
 ['HOOHOHO', 'OHOHOHO', 'HHHHOHO', 'HOHOOHO', 'HOHHHHO', 'HOHOHOO', 'HOHOHHH'])

In [89]:
with open('data2015/day19.txt') as f19:
    day19info = [row.strip() for row in f19.readlines()]
    
gap_idx = day19info.index('')
gap_idx

43

In [90]:
rep_tuples, molecule = get_rep_tuples('\n'.join(day19info[:gap_idx])), day19info[gap_idx+1]

In [91]:
molecule

'CRnCaSiRnBSiRnFArTiBPTiTiBFArPBCaSiThSiRnTiBPBPMgArCaSiRnTiMgArCaSiThCaSiRnFArRnSiRnFArTiTiBFArCaCaSiRnSiThCaCaSiRnMgArFYSiRnFYCaFArSiThCaSiThPBPTiMgArCaPRnSiAlArPBCaCaSiRnFYSiThCaRnFArArCaCaSiRnPBSiRnFArMgYCaCaCaCaSiThCaCaSiAlArCaCaSiRnPBSiAlArBCaCaCaCaSiThCaPBSiThPBPBCaSiRnFYFArSiThCaSiRnFArBCaCaSiRnFYFArSiThCaPBSiThCaSiRnPMgArRnFArPTiBCaPRnFArCaCaCaCaSiRnCaCaSiRnFYFArFArBCaSiThFArThSiThSiRnTiRnPMgArFArCaSiThCaPBCaSiRnBFArCaCaPRnCaCaPMgArSiRnFYFArCaSiThRnPBPMgAr'

In [92]:
part1 = get_distinct_new_mol(rep_tuples, molecule)
len(part1)

509

### TODO - Day 19 Part 2

<a class="anchor" id="day20"></a>
## Day 20

Elves delivering presents to an infinite row of houses.

**Part 1**  
There is an infinite row of houses numbered 1, 2, ...  
There are an infinite number of elves numbered 1, 2, ...  
In order, each elf delivers 10 presents to every number house that is an integer multiple of their elf number (so elf number 7 delivers 10 presents to every house number that is a multiple of 7).  
In this way, what is the lowest house number to receive at least as many presents as our puzzle input 34000000?  

**Part 2**  
Now, each elf delivers 11 presents to each visited house, but they only visit their first 50 houses that are multiples of their elf number.    
In this way, what is the lowest house number to receive at least as many presents as our puzzle input?

In [93]:
def get_divisors(n):
    divs = [1, n]
    for i in range(2, int(n**.5)+1):
        if n % i == 0:
            divs = divs + [i, n//i]
    return sorted(set(divs))


In [94]:
get_divisors(36)

[1, 2, 3, 4, 6, 9, 12, 18, 36]

In [95]:
get_divisors(1)

[1]

In [96]:
target = 34000000

i = 1
while True:
    presents = 10*sum(get_divisors(i))
    if presents > target:
        print(f'house number: {i}')
        break
    if i % 100000 == 0:
        print(i, presents)
    i += 1


100000 2460780
200000 4960620
300000 9843120
400000 9960300
500000 12304530
600000 19842480
700000 19686240
house number: 786240


In [97]:
present_dict = dict()
elf = 1
not_found = True
while not_found:
    for house_num in range(1, 51):
        house = elf*house_num
        if house not in present_dict:
            present_dict[house] = 11*elf
        else:
            present_dict[house] += 11*elf
    if present_dict[elf] > target:
        print(f'house number: {elf}')
        not_found = False
    if elf % 100000 == 0:
        print(elf)
    elf+= 1

100000
200000
300000
400000
500000
600000
700000
800000
house number: 831600


<a class="anchor" id="day21"></a>
## Day 21

RPG Simulator.  

We'll be simulating 1-on-1 combat with an enemy, RPG style.  The Player (us) and the enemy (the boss) all have properties of HP, Attack, Defense.  Initially, our Player starts with 100HP, 0 att, 0 def.  

Combat involves each combatant taking turns, with the Player always going first.  Attacks deal damage equal to min(1, att-def).  So, attacker has attack of 8 and defender has defense of 5, 3 damage is done, but even if the defender's defense was 10, 1 damage would still be dealt - always at least one damage.

There is a shop from which the Player can purchase items - we have as much money as we want, but our goal for **Part 1** is to determine the least amount of money we can spend at the shop and still beat the boss (whose stats are given as our puzzle input).

The Shop sells weapons, armor, and rings.  
1. We are required to buy/use exactly 1 sword. 
2. We may buy/use 0 or 1 pieces of armor. 
3. We may buy/use 0, 1, or 2 rings.


    Weapons:    Cost  Damage  Armor
    Dagger        8     4       0
    Shortsword   10     5       0
    Warhammer    25     6       0
    Longsword    40     7       0
    Greataxe     74     8       0

    Armor:      Cost  Damage  Armor
    Leather      13     0       1
    Chainmail    31     0       2
    Splintmail   53     0       3
    Bandedmail   75     0       4
    Platemail   102     0       5

    Rings:      Cost  Damage  Armor
    Damage +1    25     1       0
    Damage +2    50     2       0
    Damage +3   100     3       0
    Defense +1   20     0       1
    Defense +2   40     0       2
    Defense +3   80     0       3
    

For example, suppose you have 8 hit points, 5 damage, and 5 armor, and that the boss has 12 hit points, 7 damage, and 2 armor:

In [98]:
# = [hp, att, def]
p = [8, 5, 5]
b = [12, 7, 2]

def do_combat(p, b):
    # returns True if player wins, False if boss wins
    #print('Player\tBoss')
    while True:
        b[0] -= max([1, p[1] - b[2]])
        if b[0] <= 0:
            break
        p[0] -= max([1, b[1] - p[2]])
        if p[0] <= 0:
            break
        #print(str(p[0]) + '\t' + str(b[0]))
    #print(str(p[0]) + '\t' + str(b[0]))
    return p[0] > 0

do_combat(p, b)

True

In [99]:
import numpy as np
weapons = np.array([[8, 4, 0],
                    [10, 5, 0],
                    [25, 6, 0],
                    [40, 7, 0],
                    [74, 8, 0]])

armor = np.array([[13, 0, 1],
                  [31, 0, 2],
                  [53, 0, 3],
                  [75, 0, 4],
                  [102, 0, 5],
                  [0, 0, 0]])

rings = np.array([[25, 1, 0],
                  [50, 2, 0],
                  [100, 3, 0],
                  [20, 0, 1],
                  [40, 0, 2],
                  [80, 0, 3],
                  [0, 0, 0],
                  [0, 0, 0]])

In [100]:
weapons[1] + armor[4] + rings[1]

array([162,   7,   5])

In [101]:
len(weapons)

5

In [102]:
stat_list = []
for w in weapons:
    for a in armor:
        for i, r1 in enumerate(rings[:-1]):
            for j, r2 in enumerate(rings[i+1:]):
                stats = w + a + r1 + r2
                stat_list.append(stats)

In [103]:
stat_list.sort(key=lambda x: x[0])
stat_list[:20]

[array([8, 4, 0]),
 array([10,  5,  0]),
 array([21,  4,  1]),
 array([23,  5,  1]),
 array([25,  6,  0]),
 array([28,  4,  1]),
 array([28,  4,  1]),
 array([30,  5,  1]),
 array([30,  5,  1]),
 array([33,  5,  0]),
 array([33,  5,  0]),
 array([35,  6,  0]),
 array([35,  6,  0]),
 array([38,  6,  1]),
 array([39,  4,  2]),
 array([40,  7,  0]),
 array([41,  4,  2]),
 array([41,  4,  2]),
 array([41,  5,  2]),
 array([43,  5,  2])]

In [104]:
for stat in stat_list:
    cost, attack, defense = stat
    player = [100, attack, defense]
    boss = [104, 8, 1]
    if do_combat(player, boss):
        print(f'cheapest winner: [remaining hp, att, def] = {player}')
        print(f'cost: {cost}')
        break

cheapest winner: [remaining hp, att, def] = [2, 8, 1]
cost: 78


In [105]:
# part 2 - which is the most expensive loser?
for stat in stat_list[::-1]:
    cost, attack, defense = stat
    player = [100, attack, defense]
    boss = [104, 8, 1]
    if not do_combat(player, boss):
        print(f'most expensive loser: [final hp, att, def] = {player}')
        print(f'cost: {cost}')
        break

most expensive loser: [final hp, att, def] = [-2, 7, 2]
cost: 148


<a class="anchor" id="day22"></a>
## Day 22

Wizard Simulator.

Same combat setup as in Day 21, but now instead of items we have spells. No defensive armor (but maybe magical armor, so still a defense stat), no regular attacks, but spells have attack levels.  Spells also cost mana - we start with 500 mana, and if we can't afford to cast any spell on our turn, we lose.

Spells available:

    Magic Missile costs 53 mana. It instantly does 4 damage.
    
    Drain costs 73 mana. It instantly does 2 damage and heals you for 2 hit points.
    
    Shield costs 113 mana. It starts an effect that lasts for 6 turns. While it is active, your armor is increased by 7.
    
    Poison costs 173 mana. It starts an effect that lasts for 6 turns. At the start of each turn while it is active, it deals the boss 3 damage.
    
    Recharge costs 229 mana. It starts an effect that lasts for 5 turns. At the start of each turn while it is active, it gives you 101 new mana.
    
**Part 1**  
What is the least amount of total mana we can use in the fight and still win?

In [107]:
effects = {}
t = 1

# stats= [hp, mp, att, def]
player = [10, 250, 0, 0]
boss = [14, 0, 8, 0]

actions = ['r', 's', 'd', 'p', 'm']
# for action in actions:
#     if action == 'r':
        

In [108]:
from data2015.day22wizard import Combat

player died
boss died
won with mana cost:  0


In [109]:
p_hp, p_mp = 50, 500
b_hp, b_att = 55, 8

So, we need to deal a total of at least 55 points of damage.

In total, poison does 18, and drain does 2 and magic missile does 4.

Probably need to recursively search for all possible sequences of spells that keep mana above 0 and deal 55 damage.  Then, we do those combats with the class we made and determine which ones win.  Then we rank them by total mana costs to find the cheapest.

<a class="anchor" id="day23"></a>
## Day 23

Assembly!

We have a computer with two registers, a and b, which both start with a value of 0.  The computer supports 6 operations:

1. **hlf r** sets register r to half its current value, then continues with the next instruction.
2. **tpl r** sets register r to triple its current value, then continues with the next instruction.
3. **inc r** increments register r, adding 1 to it, then continues with the next instruction.
4. **jmp offset** is a jump; it continues with the instruction offset away relative to itself.
5. **jie r, offset** is like jmp, but only jumps if register r is even ("jump if even").
6. **jio r, offset** is like jmp, but only jumps if register r is 1 ("jump if one", not odd).

The program exits when it tries to access a non-existant instruction (beyond the length of the list).

**Part 1**  
What is the value in register b when the program (puzzle input) is finished running?

**Part 2**


In [110]:
sample = '''inc a
jio a, +2
tpl a
inc a'''

sample_insts = sample.split('\n')
sample_insts

['inc a', 'jio a, +2', 'tpl a', 'inc a']

In [111]:
class Computer:
    
    def __init__(self, a=0, b=0):
        self.a = a
        self.b = b
        self.idx = 0
        
    def do_instruction(self, inst):
        i1, i2 = inst[:3], inst[4:]

        if i1 == 'hlf':
            if i2 == 'a':
                self.a = self.a // 2
            else:
                self.b = self.b // 2
            self.idx += 1

        if i1 == 'tpl':
            if i2 == 'a':
                self.a *= 3
            else:
                self.b *= 3
            self.idx += 1

        if i1 == 'inc':
            if i2 == 'a':
                self.a += 1
            else:
                self.b += 1
            self.idx += 1
                
        if i1 == 'jmp':
            self.idx += int(i2)
            
        if i1 == 'jie':
            r, offset = i2.split(', ')
            if r == 'a':
                if self.a % 2 == 0:
                    self.idx += int(offset)
                else:
                    self.idx += 1
            else:
                if self.b % 2 == 0:
                    self.idx += int(offset)
                else:
                    self.idx += 1
            
        if i1 == 'jio':
            r, offset = i2.split(', ')
            if r == 'a':
                if self.a == 1:
                    self.idx += int(offset)
                else:
                    self.idx += 1
            else:
                if self.b == 1:
                    self.idx += int(offset)
                else:
                    self.idx += 1
                    
    def run_program(self, inst_list):
        while self.idx < len(inst_list):
            self.do_instruction(inst_list[self.idx])
        return f'a: {self.a}     b: {self.b}'

In [112]:
scomp = Computer()
scomp.run_program(sample_insts)

'a: 2     b: 0'

In [113]:
with open('data2015/day23.txt') as f23:
    instructions = [row.strip() for row in f23.readlines()]
    
instructions[-5:]

['tpl a', 'inc a', 'jmp +2', 'hlf a', 'jmp -7']

In [114]:
comp1 = Computer()
comp1.run_program(instructions)

'a: 1     b: 307'

In [115]:
comp2 = Computer(a=1, b=0)
comp2.run_program(instructions)

'a: 1     b: 160'

<a class="anchor" id="day24"></a>
## Day 24

Packing presents into the sleigh.

We're given a list of package weights, and there are three things we must consider when packing the sleigh.

First, we need to divide the packages (which are identified by weight as our puzzle input) into three groups of exactly equal weight.

We'd like to fill one compartment with as few packages as possible.

Of all possible configurations associated with having the fewest possible number of packages in a single compartment, we want to select the smallest grouping that has the smallest product of weights (its "quantum entanglement").

**Part 1**  
What is the quantum entanglement of the small group of packages in ideal configuration?

In [116]:
sample = [1, 2, 3, 4, 5, 7, 8, 9, 10, 11]
goal = sum(sample) // 3
goal

20

In [117]:
import itertools

def get_smallest_groups(weights, goal):
    results = []
    for i in range(1, len(weights)-1):
        combos = list(itertools.combinations(weights, i))
        for c in combos:
            if sum(c) == goal:
                results.append(c)
        if len(results) > 0:
            return results
    return 'no groups'

In [118]:
get_smallest_groups(sample, 20)

[(9, 11)]

In [119]:
with open('data2015/day24.txt') as f24:
    weights = [int(row.strip()) for row in f24.readlines()]
    
weights[-5:]

[101, 103, 107, 109, 113]

In [120]:
smallest_groups = get_smallest_groups(weights, sum(weights) // 3)

In [121]:
def prod(L):
    result = 1
    for x in L:
        result *= x
    return result

part1 = min([prod(group) for group in smallest_groups])
part1

10723906903

In [122]:
smallest_groups2 = get_smallest_groups(weights, sum(weights) // 4)
part2 = min([prod(group) for group in smallest_groups2])
part2

74850409

<a class="anchor" id="day25"></a>
# Day 25

Puzzle input: <mark> 20151125 </mark>

To continue, please consult the code grid in the manual.  Enter the code at row 2978, column 3083.

In [123]:
import numpy as np

grid = np.zeros(shape=(4, 4))
grid

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [124]:
grid = np.zeros(shape=(10000,10000))
i, j, n = 0, 0, 1
grid[i, j] = 20151125
grid[:5, :5]

array([[20151125.,        0.,        0.,        0.,        0.],
       [       0.,        0.,        0.,        0.,        0.],
       [       0.,        0.,        0.,        0.,        0.],
       [       0.,        0.,        0.,        0.,        0.],
       [       0.,        0.,        0.,        0.,        0.]])

In [125]:
def get_next_num(value):
    return (value * 252533) % 33554393

get_next_num(20151125)

31916031

In [126]:
# verifying the sample

num = 20151125
for _ in range(30):
    while i > 0:
        i -= 1
        j += 1
        num = get_next_num(num)
        grid[i, j] = num
    i, j = j+1, 0
    num = get_next_num(num)
    grid[i, j] = num

grid[:6, :6]

array([[20151125., 18749137., 17289845., 30943339., 10071777., 33511524.],
       [31916031., 21629792., 16929656.,  7726640., 15514188.,  4041754.],
       [16080970.,  8057251.,  1601130.,  7981243., 11661866., 16474243.],
       [24592653., 32451966., 21345942.,  9380097., 10600672., 31527494.],
       [   77061., 17552253., 28094349.,  6899651.,  9250759., 31663883.],
       [33071741.,  6796745., 25397450., 24659492.,  1534922., 27995004.]])

In [127]:
# part 1

grid = np.zeros(shape=(10000,10000))
i, j, n = 0, 0, 1
grid[i, j] = 20151125
num = 20151125

while grid[2977, 3082] == 0:
    while i > 0:
        i -= 1
        j += 1
        num = get_next_num(num)
        grid[i, j] = num
        n += 1
    i, j = j+1, 0
    num = get_next_num(num)
    grid[i, j] = num
    n += 1

grid[2977, 3082]

2650453.0

**Part 2**

not allowed until all other 49 stars are done